### PR TITLE
Fix status writer doing unnecessary update

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
Fix https://github.com/istio/istio/issues/44613

This fixes unnecessary updates when the resource conditions remain the same. 

However, it doesn't address situations where updateStatus encounters an error. This is a somewhat tricky issue in the statuswriter may need a lot more work. I tried to do with a backoff strategy, but it seems that writeFunc isn't the appropriate layer for this approach.

### CODE REVIEW

File: pilot/pkg/status/manager.go [hunk 1]
```diff
@@ -31,21 +34,38 @@ import (
...
-		scope.Debugf("writing status for resource %s/%s", m.Namespace, m.Name)
 		status := istatus.(GenerationProvider)
 		m.Status = status.Unwrap()
+		lastStatus, ok := manager.lastStatuses.Load(m.Key())
+		if !ok {
+			cfg := store.Get(m.GroupVersionKind, m.Name, m.Namespace)
+			lastStatus = cfg.Status
+			manager.lastStatuses.Store(m.Key(), lastStatus)
+		}
+		if reflect.DeepEqual(lastStatus, m.Status) {
```
Comment (howardjohn) [inline pilot/pkg/status/manager.go:56]: usually m.status is going to be a protobuf, doing reflect.DeepEqual on protobuf is not safe (will usually return false even when it shouldn't)

**Comment by hzxuzhonghu** [comment]
Do we have test covering this case

**Comment by hanxiaop** [comment]
> Do we have test covering this case

Looks like we don't. I'll update with some tests.

**Review by howardjohn — CHANGES_REQUESTED** [review]
I am not sure this is safe? This assumes we have exclusive access on the status field -- we don't. It can change at any time, and it is our responsibility to reconcile it.

Additionally, its not clear to me why we need to store kubernetes objects in a map - the informers already do that, and correctly update when other sources change the status. Our implementation already ensures we do not do NOP updates wrt to the informer state as far as I know

**Comment by hanxiaop** [comment]
> I am not sure this is safe? This assumes we have exclusive access on the status field -- we don't. It can change at any time, and it is our responsibility to reconcile it.

I was thinking we do since this is used in the old feature `PILOT_ENABLE_CONFIG_DISTRIBUTION_TRACKING` as well. Now I realized it's not since gateway is using the statuswriter as well which do not have exlusive access.

> Additionally, its not clear to me why we need to store kubernetes objects in a map - the informers already do that, and correctly update when other sources change the status. Our implementation already ensures we do not do NOP updates wrt to the informer state as far as I know

For this, same feature described above, the old feature had a ticker that frequently updated the resource, which is unrelated to the informer.

It looks like I'm not addressing the problem outlined in the issue. The fixes should now aim at optimizing the status update for the gateway and preventing NOP updates for the old feature. Also, the write operations must be safe.

I'm going to revisit the code and share my thoughts here.

**Comment by howardjohn** [comment]
Sounds good. I am a bit worried we are spending a lot of time on a feature which, IMO, is fundamentally flawed and I worry will never promote out of alpha with only incremental changes. I will bring this up in the WG

**Pull request closed without merging**